opt_einsum/backends/shared.py

```diff
@@ -0,0 +1,116 @@
+import contextlib
+import numbers
+from collections import OrderedDict
+
+from ..parser import get_symbol
+from .dispatch import get_func
+
+_SHARING_STACK = []
+_CURRENT_BACKEND = []
+
+
+@contextlib.contextmanager
+def shared_intermediates(cache=None):
+    """Context in which contract intermediate results are shared.
+
+    Note that intermediate computations will not be garbage collected until
+    1. this context exits, and
+    2. the yielded cache is garbage collected (if it was captured).
+
+    Parameters
+    ----------
+    cache : dict
+        If specified, a user-stored dict in which intermediate results will
+        be stored. This can be used to interleave sharing contexts.
+
+    Returns
+    -------
+    cache : dict
+        A dictionary in which sharing results are stored. If ignored,
+        sharing results will be garbage collected when this context is
+        exited. This dict can be passed to another context to resume
+        sharing.
+    """
+    if cache is None:
+        cache = {}
+    _SHARING_STACK.append(cache)
+    yield cache
+    _SHARING_STACK.pop()
```

Comment:
The try-finally is automatically performed by `@contextlib.contextmanager`

Reply:
Are you sure about this specifically for ``contextmanager``? The [docs](https://docs.python.org/3/library/contextlib.html) don't mention it and in my tests it is needed otherwise the post yield code is never reached. Testing with these snippets:
```python
from contextlib import contextmanager

things = []

@contextmanager
def TrySomething(value):
    things.append(value)
#     try:
    yield things[-1]
#     finally:
    things.pop() 
```
then
```python
with TrySomething('hello'):
    raise ValueError
```
gives  ``things=['hello']`` without the try/finally but ``[]`` with.

Reply:
Gosh, it looks like I'm mistaken. I'll fix and push. Thanks for catching this!